### PR TITLE
Big arrays sliced from netty buffers (long)

### PR DESCRIPTION
**Pull request by martijnvg**
Based on #90745 but for longs. This should allow aggregations down the road to read long values directly from netty buffer, rather than copying it from the netty buffer.

Relates to #89437

### CODE REVIEW

**Review by not-napoleon — COMMENTED** [review]
This looks good to me.  I think it's still good practice to get someone from Distributed to review these changes before merging though.

File: server/src/main/java/org/elasticsearch/common/bytes/CompositeBytesReference.java [hunk 1]
```diff
@@ -238,6 +238,18 @@ public int getIntLE(int index) {
         return super.getIntLE(index);
     }
 
+    @Override
+    public long getLongLE(int index) {
+        int i = getOffsetIndex(index);
+        int idx = index - offsets[i];
+        int end = idx + 8;
+        BytesReference wholeDoublesLivesHere = references[i];
```
[the remark this answers is not in the document]
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/common/bytes/CompositeBytesReference.java:246]: Yes, a classic copy paste error. I will change the variable name.

File: server/src/main/java/org/elasticsearch/common/util/BigArrays.java [hunk 2]
```diff
@@ -261,6 +261,13 @@ public void set(long index, byte[] buf, int offset, int len) {
             assert index >= 0 && index < size();
             System.arraycopy(buf, offset << 3, array, (int) index << 3, len << 3);
         }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            int size = (int) size();
+            out.writeVInt(size * 8);
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/common/util/BigArrays.java:268]: Maybe just cache `size * Long.BYTES` to a variable? :) Calculating it twice in two different ways back to back makes me unhappy :D

File: server/src/main/java/org/elasticsearch/common/util/BigArrays.java [hunk 3]
```diff
@@ -261,6 +261,13 @@ public void set(long index, byte[] buf, int offset, int len) {
             assert index >= 0 && index < size();
             System.arraycopy(buf, offset << 3, array, (int) index << 3, len << 3);
         }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            int size = (int) size();
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/common/util/BigArrays.java:267]: Maybe use `Math.toIntExcact(sizze())` here to make this obviously correct?

File: server/src/main/java/org/elasticsearch/common/util/BigDoubleArray.java [hunk 4]
```diff
@@ -134,18 +135,6 @@ public void set(long index, byte[] buf, int offset, int len) {
...
-            for (byte[] page : pages) {
-                out.write(page);
-            }
-            return;
-        }
-        for (int i = 0; i < pages.length - 1; i++) {
-            out.write(pages[i]);
-        }
-        out.write(pages[pages.length - 1], 0, lastPageEnd * Double.BYTES);
+        writePages(out, (int) size, pages, Double.BYTES, DOUBLE_PAGE_SIZE);
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/common/util/BigDoubleArray.java:138]: Maybe use `Math.toIntExcact(sizze())` here to make this obviously correct?

File: server/src/main/java/org/elasticsearch/common/util/BigLongArray.java [hunk 5]
```diff
@@ -126,4 +128,24 @@ public static long estimateRamBytes(final long size) {
     public void set(long index, byte[] buf, int offset, int len) {
         set(index, buf, offset, len, pages, 3);
     }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        writePages(out, (int) size, pages, Long.BYTES, LONG_PAGE_SIZE);
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/common/util/BigLongArray.java:134]: Maybe use `Math.toIntExcact(sizze())` here to make this obviously correct?

**Review by original-brownbear — APPROVED** [review]
LGTM just a few trivial points that might be nice to clean up :)

File: server/src/main/java/org/elasticsearch/common/util/BigArrays.java (hunk 3 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/common/util/BigArrays.java]: Thanks for the review! I pushed the changes via: ebe635c3902ead50040a8a5d879015231867f4e1

**Comment by martijnvg** [comment]
@elasticmachine run elasticsearch-ci/part-2

**Pull request merged**